### PR TITLE
Access log trailers

### LINKED ISSUE #3073: Access response trailers in access log
gRPC passes status in trailers, and it is not possible to log it, since access log interface works only with request and response headers.

The suggestion is to add some sort of `TRAIL(X?Y):Z` format rule.

### PR DESCRIPTION
**Pull request by baranov1ch**
*access log*: *response trailer logging*

*Description*:
This PR introduces new `TRAILER(X?Y):Z` formatter for file-based access logs and new fields in grpc access log service protocol, allowing to log response trailer values.

*Risk Level*: Medium

*Testing*:
Mostly unit tests. Test cases modification for access/upstream logging in router, http_connection_manager, access_loggers/http_grpc and access_log_formatter unit tests.

*Docs Changes*:
- `api/envoy/config/accesslog/v2/als.proto` - new field `HttpGrpcAccessLogConfig#additional_response_trailers_to_log` - lists response trailer names to log into grpc access log service.

- `api/envoy/config/filter/accesslog/v2/accesslog.proto` - new field `HTTPResponseProperties#response_trailers` response trailers to be logged into grpc access log service.

- Mention new logging option in `docs/root/configuration/access_log.rst`.

*Release Notes*:
access log: ability to log response trailers

[Fixes #3073]

### CODE REVIEW

**Comment by mattklein123** [comment]
@ccaraman can you take first pass on this?

File: docs/root/configuration/access_log.rst [hunk 1]
```diff
@@ -164,6 +164,13 @@ The following command operators are supported:
   TCP
     Not implemented ("-").
 
+%TRAIL(X?Y):Z%
```
Comment (mattklein123) [inline docs/root/configuration/access_log.rst]: Thoughts on just using `%TRAILER` here? It's 2 more chars and I think more clear (and more foreign than REQ and RESP).

File: include/envoy/access_log/access_log.h [hunk 2]
```diff
@@ -62,6 +62,7 @@ class Instance {
    *                      the request headers.
    */
   virtual void log(const Http::HeaderMap* request_headers, const Http::HeaderMap* response_headers,
+                   const Http::HeaderMap* response_trailers,
```
Comment (mattklein123) [inline include/envoy/access_log/access_log.h]: please update method docs, same below/elsewhere

Comment (baranov1ch) [reply include/envoy/access_log/access_log.h]: done. Seems like it was the only place though:)

File: docs/root/configuration/access_log.rst (hunk 1 above)
Comment (baranov1ch) [reply docs/root/configuration/access_log.rst]: Renamed to TRAILER. Abbreviations are very contagious :)

Comment (baranov1ch) [reply docs/root/configuration/access_log.rst]: Renamed to TRAILERS. abbreviations are so contagious.

**Pull request merged**